content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+---
+layout: post
+title: "Best Practices for Scalable Pipeline Code"
+tags:
+- pipeline
+- performance
+- scalability
+author: svanoort
+---
+====
+This by link:https://github.com/svanoort[Sam Van Oort], Software Engineer at
+link:http://cloudbees.com[CloudBees] and contributor to the Jenkins project.
+====
+
+=== Promises
+
+Today I'm going to show you best practices to write scalable and robust Jenkins pipeline. This is drawn from a
+combination of work with the internals of Pipeline and observations with high-scalability users.
+
+Pipeline code works beautifully for its intended role of orchestrating
+build/test/deploy/administer cycles.  As it is pressed into more complex roles and unanticipated uses, some users hit issues.
+In these cases, applying the best practices can make the difference between:
+
+* A single master running
+https://www.cloudbees.com/so-you-want-build-worlds-biggest-jenkins-cluster[hundreds
+of concurrent builds] on low end hardware (4 CPU cores and 4 GB of
+heap)
+* Running a couple dozen builds and bringing a master to its knees or
+crashing it...even with 16+ CPU cores and 20+ GB of heap!
+
+This has been seen in the wild.
+
+[[fundamentals]]
+Fundamentals
+------------
+
+To understand pipeline behavior you must understand a few points about
+how it executes.
+
+1.  All of the pipeline logic except for steps runs on the master (the
+Groovy). Whether simple _or_ complex! Even inside a "node" block!
+2.  _Steps_ may use executors to do work where appropriate, but each
+step has a small on-master overhead too.
+3.  Pipeline code is written as Groovy but the execution model is
+radically transformed at compile-time to Continuation Passing Style
+(CPS).
+4.  This transformation provides valuable safety and durability
+guarantees for pipelines, but it comes with trade-offs:
+    ..  Steps can invoke Java and execute fast and efficiently, but Groovy
+is _much_ slower to run than normal.
+    ..  Groovy logic requires far more memory, because an object-based
+syntax/block tree is kept in memory. +
+5.  Pipelines persist the program and its state frequently to be able to
+survive failure of the master.
+
+From these we arrive at a set of best practices to make pipelines more
+effective.
+
+[[best-practices-for-pipeline-code]]
+Best Practices For Pipeline Code
+--------------------------------
+
+. *Think of pipeline code as glue:* just enough Groovy code to connect
+together the pipeline steps and integrate tools, and no more.
+..   This makes code easier to maintain, more robust against bugs, and
+reduces load on masters.
+.  *Keep it simple:* limit the amount of complex logic embedded in the
```

Comment:
I would rater disagree with it, because Pipeline is OOP. And this OOP features are really-really useful when you build your own Pipeline libraries to create an internal automation framework. Likely it deserves a separate blogpost since I'm already working on such talk for a meetup.